### ISSUE TITLE
[BUG] Unexpected inability of compiler to optimize away moves

### ISSUE THREAD

**Issue by Brian-M-J** [modularml/mojo#2410]
### Bug description

I was testing return value optimization in Mojo when I ran into some unexpected behaviour. For the first snippet linked below, the compiler is able to optimize away the move after returning from `add_10` as demonstrated by the output:
```
Called constructor for DemoStruct(42)
Entering add_10 function
Called constructor for DemoStruct(52)
After calling add_10 function
DemoStruct(42)
Calling destructor for DemoStruct(42)
DemoStruct(52)
Calling destructor for DemoStruct(52)
```
But for the second snippet linked below, the compiler is unable to optimize away the move even though the behaviour of the function is still the same:
```
Called constructor for DemoStruct(42)
Entering add_10 function
Called constructor for DemoStruct(52)
Exiting add_10 function
Called move constructor for DemoStruct(52)
After calling add_10 function
DemoStruct(42)
Calling destructor for DemoStruct(42)
DemoStruct(52)
Calling destructor for DemoStruct(52)
```
It's better that programmers shouldn't worry about reordering statements introducing hidden costs and affecting performance.

### Steps to reproduce

Move optimized away: https://gist.github.com/modularbot/117c6679aea61ee5143ada73b3f2eb5d

Move not optimized away: https://gist.github.com/modularbot/c60ce97f28c5dcf04233aa601814c8e3

### System information

```shell
The Mojo playground as of 25-04-2024
```

**Comment by Brian-M-J** [comment]
Another place the compiler should be able to optimize away moves is for functions/methods that take their inputs as `inout` vs `owned`:

[When the argument is taken as `inout`](https://gist.github.com/modularbot/77eada9dda1ee719d3f04719deffb8b6) and mutated, no moves occur:
```
Called constructor for DemoStruct(42)
Entering modify function
Exiting modify function
Entering modify function
Exiting modify function
DemoStruct(20)
Calling destructor for DemoStruct(20)
```

[When the input is taken as `owned`](https://gist.github.com/modularbot/8409f3e22e7da946d2e84c798c27f40e) and mutated, moves do occur even though the effect of the code is the same:
```
Called constructor for DemoStruct(42)
Entering modify function
Exiting modify function
Called move constructor for DemoStruct(10)
Entering modify function
Exiting modify function
Called move constructor for DemoStruct(20)
DemoStruct(20)
Calling destructor for DemoStruct(20)
```

This introduces hidden run time costs for programmers who prefer pure functions and fluent interfaces.

Even if there is no mutation, [borrowing](https://gist.github.com/modularbot/992918b81c23a3202d6a76dff332641d) does not call `__moveinit__`:
```
Called constructor for DemoStruct(42)
Entering display function
DemoStruct(42)
Exiting display function
Entering display function
DemoStruct(42)
Exiting display function
Calling destructor for DemoStruct(42)
```

...but [passing through `owned`](https://gist.github.com/modularbot/5adaa5b685efe9f99ba973d9e8811ed1) does:
```
Called constructor for DemoStruct(42)
Entering display function
DemoStruct(42)
Exiting display function
Called move constructor for DemoStruct(42)
Entering display function
DemoStruct(42)
Exiting display function
Called move constructor for DemoStruct(42)
Calling destructor for DemoStruct(42)
```

**Comment by nmsmith** [comment]
Just FYI, Mojo is [likely going to offer](https://github.com/modularml/mojo/issues/1458) _explicit_ return value optimization, which will be useful in cases where you really, really want to make sure that a return value isn't moved.

Not that this means general improvements to code generation aren't valuable. Those are very nice as well. 🙂

**Comment by Brian-M-J** [comment]
> Just FYI, Mojo is [likely going to offer](https://github.com/modularml/mojo/issues/1458) _explicit_ return value optimization, which will be useful in cases where you really, really want to make sure that a return value isn't moved.
> 
> Not that this means general improvements to code generation aren't valuable. Those are very nice as well. 🙂

No, this is great. I much prefer the syntactically explicit, guaranteed optimizations that Mojo offers. It removes the worries of relying on the compiler being "sufficiently smart" such that it can recognize the opportunity to perform these optimizations (parallelizing, vectorizing, RVO) in all cases. No more relying on "compiler magic".

General codegen improvements like these just ensure that one doesn't have to perform a lot of syntactic ceremony to get good enough performance.

**Comment by soraros** [comment]
We have `out` convention now, this is stale.


```mojo
struct S:
  var value: Int

  fn __init__(out self, value: Int = 0):
      self.value = value
      print("init", self.value)

  fn __moveinit__(out self, owned other: Self):
      self.value = other.value
      print("move", other.value)

fn f(x: S) -> S:
  var res = S(x.value + 1)
  return res^

fn g(x: S, out res: S):
  res = S(x.value + 2)

fn main():
  var a = S()
  print("f ---")
  var f = f(a)  # The move constructor for S was called here
  print("g ---")
  var g = g(a)
```

**Comment by lattner** [comment]
great catch, thanks!
